File: 0046_permutations/solution.py
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        '''
        如果nums = [1, 2, 3]
        其全排列有: [1, 2, 3], [1, 3, 2], [2, 1, 3], [2, 3, 1], [3, 1, 2], [3, 2, 1]
        该题适合用DFS之回溯法求解从树的根节点[]开始按照深度逐个遍历
        回溯法算需要时刻关注两大变量：
            (1) 递归的终止条件(回溯的本质是递归)
            (1) 回溯过程中某时间点的当前状态
        
        下面首先对必要变量进行初始化
        depth: 当前所在位置所处树的深度
        path: 记录当前遍历过的数字序列, i.e. [1], [1, 2], [1, 2, 3]
        res: 最后输出的包含所有的全序列的数组
        ls_used: 由于全排列同一个数字不可出现2次, 所以需创建一个额外数组
                 跟踪每一个数字的使用情况, 其初始为全False 
        '''
        depth, path, res = 0, [], []
        ls_used = [False for _ in nums]

        # 初始化所有必要变量后开始从树的根节点([])进行深度遍历
        self.dfs(nums, depth, ls_used, path, res)
        return res

    def dfs(self, nums, depth, ls_used, path, res):
        # 递归终止条件: 达到树的尾部, 则将path中存储的数字加到res中
        if depth == len(nums):
            # 这里用path[:]才能取得path里面存储的值，否则path是空值
            res.append(path[:])
            return

        # 在当前节点挨个尝试所有没有被探索过的数字
        for i, used in enumerate(ls_used):
            # 跳过已经在path中出现的数字
            if used:
                continue
            # 如果该数字没有被使用, 则添加到path中, 并将数字状态改为True表示其已经被遍历
            path.append(nums[i])
            ls_used[i] = True

            # 递归: 往下一层进一步探索
            self.dfs(nums, depth + 1, ls_used, path, res)

            # 回溯到原来位置, 把path中最后新加入的弹出, 之前使用过数字现在变成未使用
            path.pop()
            ls_used[i] = False

File: 0046_permutations/solution.go
package _046_permutations

/*

为什么要回溯
我们不是找到一个排列就完事，要找出所有满足条件的排列。

递归结束时，结束的是当前的递归分支，还要去别的分支继续搜。

所以，要撤销当前的选择，回到选择前的状态，去选下一个选项，即切入下一个分支。

注意，往map添加的当前选择也要同时撤销。代表没有做这个选择。

退回来，把路走全，才能在一棵空间树中，回溯出所有的解。


为什么加入解集时，要将数组（在go中是切片）内容拷贝到一个新的数组里，再加入解集。

这个 path 变量是一个地址引用，结束当前递归，将它加入 res，后续的递归分支还要继续进行搜索，还要继续传递这个 path，这个地址引用所指向的内存空间还要继续被操作，
所以 res 中的 path 所引用的内容会被改变，这就不对，所以要拷贝一份内容，到一份新的数组里，然后放入 res，这样后续对 path 的操作，就不会影响已经放入 res 的内容。
*/

func permute(nums []int) [][]int {
	res := [][]int{}
	visited := map[int]bool{} // 记录选过的数，下次遇到相同的数，跳过

	var dfs func(path []int)
	dfs = func(path []int) {
		if len(path) == len(nums) {
			temp := make([]int, len(path))
			copy(temp, path)
			res = append(res, temp)
			return
		}
		for _, n := range nums {
			if visited[n] { // 不能重复选
				continue
			}
			path = append(path, n)
			visited[n] = true
			dfs(path)
			path = path[:len(path)-1]
			visited[n] = false
		}
	}

	dfs([]int{})
	return res
}
